File: day13/packet_scanners.py
#!/usr/bin/env python3
from sys import argv
import re

# parse the input
fw = []
filename = argv[1]
with open(filename) as f:
    for line in f:
        line = line.rstrip()
        m = re.search('(\d+): (\d+)', line)
        layer = int(m.group(1))
        for _ in range(len(fw), layer):
            fw.append(None)
        fw.append(int(m.group(2)))

# compute severity
severity = 0
for i in range(len(fw)):
    if fw[i] is not None and i % (2 * (fw[i]-1)) == 0:
        severity += i * fw[i]

print("result1:", severity)

# now let's try to get through without getting caught!
delay = 0
ok = False
while (not ok):
    ok = True
    for i in range(len(fw)):
        if fw[i] is not None and (delay + i) % (2 * (fw[i]-1)) == 0:
            ok = False
            break
    if not ok:
        delay += 1

print("result2:", delay)

File: day13/packet_scanners.pl
#!/usr/bin/env perl
use strict;
use warnings;
use feature qw(:5.24);
use experimental qw(signatures);

my @fw;

# parse the input
while (<>) {
    chomp;
    /(\d+): (\d+)/;
    $fw[$1] = $2;
}

# compute severity
my $severity = 0;
for my $i (0..$#fw) {
    $severity += $i * $fw[$i] if caught(\@fw, $i)
}

say "result1: $severity";

# now let's try to get through without getting caught!
my $delay = 0;
my $ok = 0;
while (!$ok) {
    $ok = 1;
    for my $i (0..$#fw) {
        if (caught(\@fw, $i, $delay)) {
            $ok = 0;
            last;
        }
    }
    $delay++ unless $ok;
}

say "result2: $delay";

sub caught($fw, $i, $delay = 0) {
    return (defined $fw->[$i] && ($delay+$i) % (2 * ($fw->[$i]-1)) == 0);
}
